lib/segment/src/id_tracker/immutable_id_tracker.rs

```diff
@@ -0,0 +1,532 @@
+        mappings: PointMappings,
+    ) -> OperationResult<Self> {
+        // Create mmap file for deleted bitvec
+        let deleted_filepath = Self::deleted_file_path(path);
+        {
+            let deleted_size = bitmap_mmap_size(deleted);
+            create_and_ensure_length(&deleted_filepath, deleted_size)?;
+        }
+
+        let mut deleted_new = MmapBitSlice::try_from(open_write_mmap(&deleted_filepath)?, 0)?;
+        deleted_new[..deleted.len()].copy_from_bitslice(deleted);
+        let deleted_wrapper = MmapBitSliceBufferedUpdateWrapper::new(deleted_new);
+
+        // Create mmap file for internal-to-version list
+        let version_filepath = Self::version_mapping_file_path(path);
+        {
+            let version_size = size_of_val(internal_to_version);
+            create_and_ensure_length(&version_filepath, version_size)?;
+        }
+        let mut internal_to_version_wrapper =
+            unsafe { MmapSlice::try_from(open_write_mmap(&version_filepath)?)? };
+        internal_to_version_wrapper.copy_from_slice(internal_to_version);
+        let internal_to_version = internal_to_version_wrapper.to_vec();
+        let internal_to_version_wrapper =
+            MmapSliceBufferedUpdateWrapper::new(internal_to_version_wrapper);
+
+        // Serialize the mappings with bincode and write them to disk
+        Self::serialize_mappings(&Self::mappings_file_path(path), &mappings)?;
+
+        Ok(Self {
+            path: path.to_path_buf(),
+            deleted: deleted.to_bitvec(),
+            deleted_wrapper,
+            internal_to_version_wrapper,
+            internal_to_version,
+            mappings,
+        })
+    }
+
+    fn deleted_file_path(base: &Path) -> PathBuf {
+        base.join(DELETED_FILE_NAME)
+    }
+
+    fn version_mapping_file_path(base: &Path) -> PathBuf {
+        base.join(VERSION_MAPPING_FILE_NAME)
+    }
+
+    pub(crate) fn mappings_file_path(base: &Path) -> PathBuf {
+        base.join(MAPPINGS_FILE_NAME)
+    }
+
+    fn serialize_mappings(path: &Path, mappings: &PointMappings) -> OperationResult<()> {
+        let mappings: Vec<_> = mappings
+            .internal_to_external
+            .iter()
+            .map(|external| {
+                let internal_id = match external {
+                    StoredPointId::NumId(n) => mappings.external_to_internal_num.get(n),
+                    StoredPointId::Uuid(u) => mappings.external_to_internal_uuid.get(u),
+                    StoredPointId::String(str) => {
+                        unimplemented!("cannot convert internal string id '{str}' to external id")
+                    }
+                }
+                .unwrap();
+                (external, internal_id)
+            })
+            .collect();
+        Self::serialize_file(path, &mappings)
+    }
+
+    fn load_mappings(path: &Path) -> OperationResult<PointMappings> {
+        let mappings: Vec<(StoredPointId, PointOffsetType)> = Self::deserialize_file(path)?;
```

Comment:
I think at some scale this can be a bottleneck, when we reading all points in memory and then build data structure.

Consider 1B points (something we ideally want to run on a single machine, assuming we have MMAP for everything). 

For the PointsMapping structure it would be `1_000_000_000 x "(24 + 16 + 4)" / 1024 / 1024 / 1024 ~= 40Gb RAM`

If we need double of than to just load from disk, that might be just to expensive. 
Potential downside - reading one-by-one might be slower, cause we would need to call deserialize function for each point (not sure how significant would it be)


Probably need a second opinion @timvisee

Reply:
I agree. We could also try to use a custom deserializer, that directly builds the required `PointMappings` Struct in [visit_seq](https://docs.rs/serde/1.0.203/serde/de/trait.Visitor.html#method.visit_seq) so we don't allocate a `Vec` first. I'd need to test and verify whether this really is faster and doesn't allocate too much memory but I can imagine this to work.

Reply:
TL;DR: This approach is ~15% faster and ~2.5x more memory efficient.

I've implemented and analyzed this approach which confirms that the custom deserializer uses way less memory. Furthermore I tested execution time (point count: 763000 and 3203400) and the new approach is ~15%  faster as well.
I'm not 100% sure if the amount of memory reported on the screenshots are only from loading itself because there seems to be some overhead of the profiler but with those metrics the new approach uses ~2.5 times less memory. This also confirms that there is no other large allocation internally.

I think this is the most performant approach that can be done with serde at the moment. The next step would be to do it our own and write/read the bytes directly into/from the file.
If you're interested in that I can sure try it out and measure its performance.

![image](https://github.com/qdrant/qdrant/assets/15957865/d786ac32-c3ff-4bd3-b728-8fb2530d27e1)
![image](https://github.com/qdrant/qdrant/assets/15957865/5f42f341-403d-4bc9-9e94-868b6cddade2)

Reply:
Nice! We're also interested in less allocations too, but I guess that's an automatic side effect of this approach.

I haven't inspected the code around this yet, but the results look promising.